controllers/machinedeletionremediation_controller.go

```diff
@@ -462,76 +483,118 @@ func (r *MachineDeletionRemediationReconciler) isTimedOutByNHC(remediation *v1al
+	allNodes := new(v1.NodeList)
+	if err := r.List(context.Background(), allNodes); err != nil {
+		return nil, err
 	}
-	if !exists {
-		log.Info("machine does not have phase", "machine", machine.GetName())
-		return "unknown"
+
+	machineSetNodes := &v1.NodeList{}
+	for _, node := range allNodes.Items {
+		machineName, machineNs, err := getMachineNameNsFromNode(&node)
+		if err != nil {
+			r.Log.Error(err, "could not get MachineNameNS from Node", "node", node.Name)
+			continue
+		}
+
+		machine, found := visitedMachines[machineName]
+		if !found {
+			key := client.ObjectKey{
+				Name:      machineName,
+				Namespace: machineNs,
+			}
+			if err := r.Get(context.TODO(), key, &machine); err != nil {
+				r.Log.Error(err, "could not get Machine of Node", "node", node.Name, "search key", key)
+				continue
+			}
+			visitedMachines[machineName] = machine
+		}
+
+		if thisMachineSetName, err := getMachineOwnerReferenceName(&machine); err != nil {
+			r.Log.Error(err, "no valid ownerReference", "node", node.Name, "machine", machine.Name)
+		} else if thisMachineSetName == machineSetName {
+			machineSetNodes.Items = append(machineSetNodes.Items, node)
+		}
 	}
-	return phase
-}
 
-func getMachineProviderID(machine *unstructured.Unstructured) (string, error) {
-	log := ctrl.Log.WithName("getMachineStatusProviderID")
+	return machineSetNodes, nil
+}
 
-	status, exists, err := unstructured.NestedMap(machine.Object, "spec")
-	if err != nil {
-		log.Error(err, "could not get machine's spec", "machine", machine.GetName())
-		return "", err
+func getMachineNameNsFromNode(node *v1.Node) (string, string, error) {
+	var nodeAnnotations map[string]string
+	if nodeAnnotations = node.Annotations; nodeAnnotations == nil {
+		return "", "", fmt.Errorf(noAnnotationsError, node.Name)
 	}
-	if !exists {
-		log.Info("machine does not have spec", "machine", machine.GetName())
-		return "", nil
+
+	var machineNameNs string
+	if machineNameNs = nodeAnnotations[machineAnnotationOpenshift]; len(machineNameNs) == 0 {
+		return "", "", fmt.Errorf(noMachineAnnotationError, node.Name)
 	}
 
-	providerID, exists, err := unstructured.NestedString(status, "providerID")
-	if err != nil {
-		log.Error(err, "could not get machine's providerID", "machine", machine.GetName())
-		return "", err
+	if slice := strings.Split(machineNameNs, "/"); len(slice) == 2 {
+		return slice[1], slice[0], nil
 	}
-	if !exists {
-		log.Info("machine does not have providerID", "machine", machine.GetName())
+	return "", "", fmt.Errorf(invalidValueMachineAnnotationError, node.Name)
+}
+
+// getMachineOwner returns the Machine's ownerReference name. It returns an error if the Machine has no ownerReference
+// or more than one.
+func getMachineOwnerReferenceName(machine *v1beta1.Machine) (string, error) {
+	owners := machine.GetOwnerReferences()
+	switch len(owners) {
+	case 0:
+		// not an error, e.g. control-plane nodes in BareMetal clusters
```

Comment:
Can you elaborate why it is not an error?

Reply:
`control-plane/master` nodes do not have an `ownerReference`. This is also why MDR skips `control-planes` remediation.